src/renderer/components/commands-editors.tsx

```diff
@@ -103,6 +150,72 @@ export class EditorDropdown extends React.Component<
     return result;
   }
 
+  public async addCustomEditor() {
+    const { appState } = this.props;
+
+    const isValidEditorName = (name: string) =>
+      /^.*\.(js|html|css)$/gm.test(name);
+
+    // Reset potentially non-null last description.
+    appState.genericDialogLastInput = null;
+
+    appState.setGenericDialogOptions({
+      type: GenericDialogType.confirm,
+      label: 'Enter a filename for your custom editor',
+      wantsInput: true,
+      ok: 'Create',
+      cancel: 'Cancel',
+      placeholder: 'file.js',
+    });
+
+    appState.isGenericDialogShowing = true;
+    await when(() => !appState.isGenericDialogShowing);
+
+    const cancelled = !appState.genericDialogLastResult;
+    if (!cancelled) {
+      const mosaicName = appState.genericDialogLastInput;
+
+      // Fail if editor name is not an accepted file type.
+      if (!mosaicName || !isValidEditorName(mosaicName)) {
+        appState.setGenericDialogOptions({
+          type: GenericDialogType.warning,
+          label:
+            'Invalid custom editor name - must be either an html, js, or css file.',
+          cancel: undefined,
+        });
+
+        appState.toggleGenericDialog();
+      } else {
+        const name = mosaicName as CustomEditorId;
+
+        // Also fail if the user tries to create two identical editors.
+        if (
+          appState.customMosaics.includes(name) ||
+          Object.values(DefaultEditorId).includes(name as DefaultEditorId)
+        ) {
+          appState.setGenericDialogOptions({
+            type: GenericDialogType.warning,
+            label: `Custom editor name ${name} already exists - duplicates are not allowed`,
+            cancel: undefined,
+          });
+
+          appState.toggleGenericDialog();
+        } else {
+          appState.customMosaics.push(name);
+          appState.showMosaic(name);
+        }
+      }
+    }
+  }
```

Comment:
this nested if... if... else... if... else... logic is kind of difficult to read and is nested pretty deep. Could it be simplified with returns, e.g. `if (cancelled) return;` and similar for the other conditionals

Reply:
As a followup: coverage of src/renderer/components/commands-editors.tsx drops 41% in this PR as of b776cb04532cdbece4a3889dceb347bf581a6c81